File: JavaSolver/src/test/java/MostCommonWord.java
import org.junit.jupiter.api.Test;

import java.util.*;
import java.util.stream.Collectors;

public class MostCommonWord {

    @Test
    public void run() {
        // String paragraph = "Bob hit a ball, the hit BALL flew far after it was hit by bAll.";
        String paragraph = "a, a, a, a, b,b,b,c, c";
        String[] banned = new String[] {"a"};

        System.out.println(solution(paragraph, banned));
    }

    private String solution(String paragraph, String[] banned) {
        String filtered = paragraph.replaceAll("[^\\w]", " ").replaceAll("\\s+", " ").toLowerCase();
        System.out.println(filtered);
        List<String> list = Arrays.stream(filtered.split(" "))
                .filter(w -> {
                    for(String ban : banned) {
                        if (w.equals(ban)) {
                            return false;
                        }
                    }
                    return true;
                }).collect(Collectors.toList());

        Map<String, Integer> counts = new HashMap<>();

        list.forEach(w -> {
            int count = counts.getOrDefault(w, 0);
            count++;
            counts.put(w, count);
        });

        return Collections
                .max(counts.entrySet(), Comparator.comparingInt(Map.Entry::getValue))
                .getKey();
    }
}

File: PySolver/most_common_word.py
import collections
import re
from typing import List


class Solution:
    def most_common_word(self, paragraph: str, banned: List[str]):
        words = [word for word in re.sub(r'[^\w]', ' ', paragraph).lower().split() if word not in banned]

        # counts = collections.defaultdict(int)
        # for word in words:
        #     counts[word] += 1
        #
        # return max(counts, key=counts.get)

        counts = collections.Counter(words)
        return counts.most_common(1)[0][0]


if __name__ == '__main__':
    paragraph = 'Bob hit a ball, the hit BALL flew far after it was hit.'
    banned = ['hit']

    print(Solution().most_common_word(paragraph, banned))
